### ISSUE TITLE
Initialized Rancher HA Stack won't complete recovery from reboot

### ISSUE THREAD

**Issue by wazoo** [rancher/rancher#4699]
Building on the conversation in #4634, I had a configured HA stack and rebooted all 3 hosts at relatively the same time for patching and now the cluster has lost its mind and refuses to come back up.

Rancher Version: 1.0.1
Docker Version:
```
# docker version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64
```
OS:
``` 
# cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.3 LTS"
# uname -a
Linux ch01-core 3.19.0-25-generic #26~14.04.1-Ubuntu SMP Fri Jul 24 21:16:20 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
# mysql -V 
mysql  Ver 14.14 Distrib 5.5.49, for debian-linux-gnu (x86_64) using readline 6.3
```
Steps to Reproduce: 
1. Start rancher with external MySQL Database
2. Run HA wizard
3. Stop bootstrap container
4. Run HA script on 3 hosts
5. Reboot all 3 hosts at the same time

Results: 
1. Rancher explodes and nothing works.

Expected:
1. Rancher should not explode and everything should work.

The cluster seems to have recovered partially over night but I noticed this morning that the UI says that there are two containers in the websocket-proxy stack that are "starting" but I see no evidence of them on my host and they don't have container IDs, same goes for go-machine ones.  They all seem to be stuck in the Starting (networking) state which is weird.  

One other wierd thing that I noticed is that in the System HA environment I had a duplicate entry for one of my hosts for some reason and the problem containers seemed to all be on one of the duplicates so I tried removing it and it actually removed both duplicates and I can't login after my session timed out so I tried restarting the services and now I am back to it not starting at all with this in the rancher-ha logs again:
```
time="2016-05-05T17:20:23Z" level=info msg="Waiting for server to be available" component=cert 
time="2016-05-05T17:20:23Z" level=info msg="Can not launch agent right now: Server not available at http://172.17.0.1:18080/ping:" component=service 
```
Checking on ZK (before removing all of the containers):
```
~# docker exec -it rancher-ha-zk bash
root@3973075a9825:/# echo mntr | nc localhost 2181
zk_version      3.4.5--1, built on 06/10/2013 17:26 GMT
zk_avg_latency  6
zk_max_latency  150
zk_min_latency  0
zk_packets_received     171836
zk_packets_sent 175853
zk_num_alive_connections        3
zk_outstanding_requests 0
zk_server_state follower
zk_znode_count  2414
zk_watch_count  58
zk_ephemerals_count     66
zk_approximate_data_size        87699
zk_open_file_descriptor_count   39
zk_max_file_descriptor_count    1048576
root@3973075a9825:/# echo mntr | nc localhost 2182
zk_version      3.4.5--1, built on 06/10/2013 17:26 GMT
zk_avg_latency  0
zk_max_latency  0
zk_min_latency  0
zk_packets_received     1
zk_packets_sent 0
zk_num_alive_connections        1
zk_outstanding_requests 0
zk_server_state leader
zk_znode_count  2450
zk_watch_count  0
zk_ephemerals_count     66
zk_approximate_data_size        89249
zk_open_file_descriptor_count   39
zk_max_file_descriptor_count    1048576
zk_followers    2
zk_synced_followers     2
zk_pending_syncs        0
root@3973075a9825:/# echo mntr | nc localhost 2183
zk_version      3.4.5--1, built on 06/10/2013 17:26 GMT
zk_avg_latency  7
zk_max_latency  162
zk_min_latency  0
zk_packets_received     18001
zk_packets_sent 18372
zk_num_alive_connections        2
zk_outstanding_requests 0
zk_server_state follower
zk_znode_count  2450
zk_watch_count  37
zk_ephemerals_count     66
zk_approximate_data_size        89249
zk_open_file_descriptor_count   38
zk_max_file_descriptor_count    1048576
root@3973075a9825:/#
```
That seems to be working at least.

Checking on rancher-agent I see this over and over at the end of the logs (before removing all of the containers):
```
INFO: Starting agent for 0610D55C3EDEFD220EAB
INFO: Access Key: 0610D55C3EDEFD220EAB
INFO: Config URL: http://172.17.0.1:18080/v1
INFO: Storage URL: http://172.17.0.1:18080/v1
INFO: API URL: http://172.17.0.1:18080/v1
INFO: IP: <ip>
INFO: Port:
INFO: Required Image: rancher/agent:v1.0.1
INFO: Current Image: rancher/agent:v1.0.1
INFO: Using image rancher/agent:v1.0.1
INFO: Downloading agent http://172.17.0.1:18080/v1/configcontent/configscripts
{"id":"4916b3a7-9b04-4d15-944c-c43243a4f3b8","type":"error","links":{},"actions":{},"status":401,"code":"Unauthorized","message":"Unauthorized","detail":null}
gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

To summarize, I have tried the following:
- Removed /var/lib/rancher/state (but didn't drop my database) and restarted, same results.
- Removed all of the containers and re-ran the HA script, same results
- Removed all of the containers and volumes and re-ran the HA script, same results

**Comment by LLParse** [comment]
So your environment is this:
ESX 6.0 On Premise
Ubuntu 14.04.3
Docker 1.10.3
Rancher 1.0.1
MySQL

What are you using for load balancing? Could we get logs for rancher-ha-cattle (that is the container that runs Rancher server)?

**Comment by wazoo** [comment]
Previously I had the URL pointed at the first node but now I am just using nginx to load balance 80 and 443 right now, here is the config:
```
upstream rancher-cluster {
        server host01:443;
        server host02:443;
        server host03:443;
}

server {
        listen 80 default;
        server_name <cluster url>;
        rewrite ^ https://<cluster url> permanent;
}

server {
    listen       443 ssl;
    server_name  <cluster url>;

    ssl_certificate      /etc/ssl/<cluster url>crt.pem;
    ssl_certificate_key  /etc/ssl/<cluster url>.key.pem;
    ssl_trusted_certificate /etc/ssl/<cluster url>-ca-chain.pem;

    location / {
        proxy_pass https://rancher-cluster;
    }
}
```
I dropped the database and re-created the cluster with a new config script to make sure that it wasn't me trying to launch two agent nodes on these hosts (I didn't notice that the HA script creates the System HA environment at the time) that broke it and it still seemes stuck on "Initializing" the rancher-compose-executor and go-machine-service but only on the first host (the one that bootstrap was run on):

```
time="2016-05-05T20:08:01Z" level=info msg="Starting go-machine-service..." gitcommit=dc97268 
time="2016-05-05T20:08:01Z" level=info msg="Updating docker-machine-drivers from cattle." 
time="2016-05-05T20:05:42Z" level=error msg="Error from driver: Get https://<cluster-url>/v1: x509: certificate signed by unknown authority" 
time="2016-05-05T20:05:42Z" level=fatal msg="Encourted an error while updating drivers." 
time="2016-05-05T20:05:42Z" level=fatal msg="exit status 1"
```

```
time="2016-05-05T20:07:15Z" level=info msg="Starting rancher-compose-executor" version=v0.7.4 
time="2016-05-05T20:07:15Z" level=fatal msg="Unable to create event router" error="Get https://<cluster url>/v1: x509: certificate signed by unknown authority" 
time="2016-05-05T20:07:15Z" level=fatal msg="exit status 1"
```

The weirdest thing is that each of these services has a "working" container on the same host (host01 lets call it) and one is running but the other is in a loop complaining about certificates.

Regardless, this issue is about it not being able to withstand all 3 nodes restarting at the same time so after it got into this "mostly working" state I restarted all 3 hosts (actual reboot to simulate outage).

From the rancher-ha-cattle I noticed this:
```
2016-05-05 20:15:38,560 ERROR [:] [] [] [] [erviceReplay-11] [i.c.p.e.e.i.ProcessEventListenerImpl] Unknown exception running process [instance.start:5268] on [38] io.cattle.platform.eventing.exception.EventExecutionException: 500 Server Error: Internal Server Error ("failed to create endpoint c2ae9705-3706-464d-9d2c-6250cff231fe on network bridge: iptables failed: iptables --wait -t filter -A DOCKER ! -i docker0 -o docker0 -p udp -d 172.17.0.3 --dport 4500 -j ACCEPT: iptables: No chain/target/match by that name.
 (exit status 1)")
```
so I stopped docker, deleted the bridge and restarted docker and now I get this (ignoring redis and zk errors while those were still starting:

### rancher-ha-cattle
```
2016-05-05 20:18:52,423 WARN  [:] [] [] [] [ecutorService-6] [org.apache.curator.utils.ZKPaths    ] The version of ZooKeeper being used doesn't support Container nodes. CreateMode.PERSISTENT will be used instead. 
2016-05-05 20:18:53,554 INFO    [main] [ConsoleStatus] [22/32] [64484ms] [51073ms] Starting system-services 
2016-05-05 20:18:53,554 INFO    [main] [ConsoleStatus] [23/32] [64484ms] [0ms] Starting agent-server 
2016-05-05 20:18:53,554 INFO    [main] [ConsoleStatus] [24/32] [64484ms] [0ms] Starting allocator-server 
2016-05-05 20:18:54,207 INFO    [main] [ConsoleStatus] [25/32] [65137ms] [653ms] Starting api-server 
2016-05-05 20:18:56,002 INFO    [main] [ConsoleStatus] [26/32] [66932ms] [1795ms] Starting iaas-api 
2016-05-05 20:18:56,002 INFO    [main] [ConsoleStatus] [27/32] [66932ms] [0ms] Starting archaius 
2016-05-05 20:18:56,002 INFO    [main] [ConsoleStatus] [28/32] [66932ms] [0ms] Starting core-model 
2016-05-05 20:18:56,002 INFO    [main] [ConsoleStatus] [29/32] [66932ms] [0ms] Starting core-object-defaults 
2016-05-05 20:18:56,002 INFO    [main] [ConsoleStatus] [30/32] [66932ms] [0ms] Starting encryption 
2016-05-05 20:18:56,219 INFO    [main] [ConsoleStatus] [31/32] [67149ms] [217ms] Starting process 
2016-05-05 20:18:56,219 INFO    [main] [ConsoleStatus] [32/32] [67149ms] [0ms] Starting redis 
20:18:56.315 [main] INFO  ConsoleStatus - [DONE ] [70582ms] Startup Succeeded, Listening on port 8080
time="2016-05-05T20:18:58Z" level=info msg="Starting Rancher Catalog service" 
time="2016-05-05T20:18:58Z" level=info msg="Using catalog library=https://github.com/rancher/rancher-catalog.git" 
time="2016-05-05T20:18:58Z" level=info msg="Using catalog community=https://github.com/rancher/community-catalog.git" 
2016-05-05 20:19:29,139 ERROR [:] [] [] [] [ecutorService-7] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [2] count [3] 
2016-05-05 20:19:29,139 ERROR [:] [] [] [] [ecutorService-7] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [3] count [3] 
2016-05-05 20:19:34,141 ERROR [:] [] [] [] [ecutorService-8] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [2] count [4] 
2016-05-05 20:19:34,141 ERROR [:] [] [] [] [ecutorService-8] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [3] count [4] 
2016-05-05 20:19:39,141 ERROR [:] [] [] [] [ecutorService-3] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [2] count [5] 
2016-05-05 20:19:39,141 ERROR [:] [] [] [] [ecutorService-9] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [3] count [5] 
2016-05-05 20:19:44,145 ERROR [:] [] [] [] [ecutorService-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [3] count [6] 
2016-05-05 20:19:44,145 ERROR [:] [] [] [] [ecutorService-3] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [2] count [6] 
2016-05-05 20:19:44,147 ERROR [:] [] [] [] [ecutorService-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Scheduling reconnect for [3] 
2016-05-05 20:19:44,147 ERROR [:] [] [] [] [ecutorService-3] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Scheduling reconnect for [2] 
2016-05-05 20:19:49,148 ERROR [:] [] [] [] [ecutorService-3] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [2] count [7] 
2016-05-05 20:19:49,148 ERROR [:] [] [] [] [ecutorService-4] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [3] count [7] 
2016-05-05 20:19:54,151 ERROR [:] [] [] [] [ecutorService-8] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [2] count [8] 
2016-05-05 20:19:54,152 ERROR [:] [] [] [] [ecutorService-7] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [3] count [8] 
2016-05-05 20:19:59,153 ERROR [:] [] [] [] [ecutorService-5] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [2] count [9] 
2016-05-05 20:19:59,153 ERROR [:] [] [] [] [ecutorService-6] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [3] count [9]
```
### rancher-ha
```
time="2016-05-05T20:18:54Z" level=info msg="Waiting for server to be available" component=cert 
time="2016-05-05T20:18:54Z" level=info msg="Can not launch agent right now: Server not available at http://172.17.0.1:18080/ping:" component=service 
```

### rancher-ha-zk
```
2016-05-05 20:22:16,816 - ERROR [LearnerHandler-/127.0.0.1:39288:LearnerHandler@562] - Unexpected exception causing shutdown while sock still open
java.io.EOFException
        at java.io.DataInputStream.readInt(DataInputStream.java:392)
        at org.apache.jute.BinaryInputArchive.readInt(BinaryInputArchive.java:63)
        at org.apache.zookeeper.server.quorum.QuorumPacket.deserialize(QuorumPacket.java:83)
        at org.apache.jute.BinaryInputArchive.readRecord(BinaryInputArchive.java:108)
        at org.apache.zookeeper.server.quorum.LearnerHandler.run(LearnerHandler.java:290)
2016-05-05 20:22:16,816 - WARN  [LearnerHandler-/127.0.0.1:39288:LearnerHandler@575] - ******* GOODBYE /127.0.0.1:39288 ********
```
A manual restart of rancher-ha-zk resulted in it starting from what I could tell but rancher-ha-cattle is still complaining with those same errors above.

**Comment by wazoo** [comment]
From what i can observe it seems like ZK is stuck in some sort of split-brain as none of the nodes are recovering.  I am going to have to just start over and paste a virtual "don't reboot these all at once" sticker on the cluster but since its easy to reproduce this issue you probably won't need any of my logs anyway.

**Comment by LLParse** [comment]
@wazoo Could you verify that your certs are on all machines in the `/var/lib/rancher/etc/ssl` folder? That seems to be the culprit of your go-machine-service and rancher-compose-executor problems (for the moment). Once you fix that, you'll probably see GMS/RCE flapping for other reasons.

I assume you've configured your `Host Registration URL` to point to your Nginx load balancer. You'll need to make a few more adjustments to the LB config to inject necessary HTTP headers. This is necessary to support websockets, which are used for shell access and log tailing. There's an [example](http://docs.rancher.com/rancher/latest/en/installing-rancher/installing-server/basic-ssl-config/) configuration for Nginx that should help clarify.

**Comment by wazoo** [comment]
Okay, I finally got a chance to get back to this.  My nginx config is now this:

```
upstream rancher-cluster {
        ip_hash;
        server ch01.fqdn:443 max_fails=3  fail_timeout=60s;
        server ch02.fqdn:443 max_fails=3  fail_timeout=60s;
        server ch03.fqdn:443 max_fails=3  fail_timeout=60s;
}

server {
        listen 80 default;
        server_name <cluster url>;
        return 301 https://$server_name$request_uri;
}

server {
    listen       443 ssl;
    server_name  <cluster url>;

    ssl_certificate      /etc/ssl/<cluster url>.crt.pem;
    ssl_certificate_key  /etc/ssl/<cluster url>.key.pem;

    location / {
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_pass https://rancher-cluster;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_read_timeout 900s;
    }
}
```
With that config it seems that rancher server boots fine and HA enables correctly however now when I try to add a kube host to my environment I get this from rancher-agent:
```
time="2016-05-27T16:07:23Z" level="info" msg="Starting event router."
time="2016-05-27T16:07:23Z" level="info" msg="Watching state directory: /var/lib/rancher/state/containers"
time="2016-05-27T16:07:23Z" level="info" msg="Processing event: &docker.APIEvents{Status:\"start\", ID:\"<id>\", From:\"-simulated-\", Time:0}"
time="2016-05-27T16:07:23Z" level="info" msg="Connecting to proxy." url="wss://<cluster url>/v1/connectbackend?token=<token>"
time="2016-05-27T16:07:23Z" level="fatal" msg="Failed to connect to proxy." error="websocket: bad handshake"
```
One change that I made was using a signed cert for <cluster url> so I don't have to worry about dealing with self signed certs.  From what I found on google my nginx config should be fine (it is basically the example you have just with iphash) but rancher-agent is just looping with that error, not sure why.  

If this needs to be a different issue I can open one but I never had trouble setting this up until I was using HA behind nginx so I think it might be related.

**Comment by wazoo** [comment]
Actually, it seems that some of my rancher-ha nodes weren't starting right so that is the cause of the error, working fine now it seems.

**Comment by LLParse** [comment]
Excellent! If you experience any other odd behavior, feel free to open another issue.
